pkg/config/parse.go

```diff
@@ -0,0 +1,106 @@
+package config
+
+import (
+	"runtime"
+
+	"github.com/spf13/viper"
+	validator "gopkg.in/go-playground/validator.v9"
+)
+
+// Config provides configuration values for all components
+type Config struct {
+	EnableCSRFProtection bool
+	GoEnv                string         `validate:"required"`
+	GoBinary             string         `validate:"required"`
+	LogLevel             string         `validate:"required"`
+	MaxConcurrency       int            `validate:"required"`
+	MaxWorkerFails       uint           `validate:"required"`
+	CloudRuntime         string         `validate:"required"`
+	FilterFile           string         `validate:"required"`
+	TimeoutSeconds       int            `validate:"required"`
+	Proxy                *ProxyConfig   `validate:""`
+	Olympus              *OlympusConfig `validate:""`
+	Storage              *StorageConfig `validate:""`
+}
+
+// ParseConfig parses the given file into an athens config struct
+func ParseConfig(configFile string) (*Config, error) {
+
+	viper.SetConfigFile(configFile)
+
+	// attempt to parse the given config file
+	if err := viper.ReadInConfig(); err != nil {
+		return nil, err
+	}
+
+	// set default values and environment variable bindings
+	setDefaultsAndEnvBindings()
+	var config Config
+	if err := viper.Unmarshal(&config); err != nil {
+		return nil, err
+	}
+	// validate all required fields have been populated
+	validate := validator.New()
+	if err := validate.Struct(config); err != nil {
+		return nil, err
+	}
+	return &config, nil
+}
+
+func setDefaultsAndEnvBindings() {
```

Comment:
This function might be hard to maintain, why don't we use `envconfig` to make our struct more declarative. Something like this: 

```golang
type Config struct {
  EnableCSRFProtection bool
  GoEnv                string         `validate:"required",envconfig:"GO_ENV",default:"development"`
  GoBinary string `validate:"required" envconfig: "GO_BINARY_PATH" default: "go"
}
```
If viper provides the same functionality, then we don't need envconfig. 

And in the initializer, you basically just do 3 steps in a row: 
1. marshal TOML into struct
2. parse envconfig (override TOML if env var exists)
3. validate struct

Reply:
I agree, it doesn't look like viper provides something like this so we should look into `envconfig`. Viper just uses [pelletier/go-toml](github.com/pelletier/go-toml) to load a map of values.